src/mpid/ch4/netmod/ofi/util.c

```diff
@@ -60,51 +62,99 @@ int MPIDI_OFI_mr_key_allocator_init(void)
-            return i * sizeof(uint64_t) * 8 + (nval - 1);
-        }
-        if (i == mr_key_allocator.num_ints - 1) {
-            mr_key_allocator.num_ints += mr_key_allocator.chunk_size;
-            mr_key_allocator.bitmask = MPL_realloc(mr_key_allocator.bitmask,
-                                                   sizeof(uint64_t) * mr_key_allocator.num_ints,
-                                                   MPL_MEM_RMA);
-            MPIR_Assert(mr_key_allocator.bitmask);
-            memset(&mr_key_allocator.bitmask[i + 1], 0xFF,
-                   sizeof(uint64_t) * mr_key_allocator.chunk_size);
-        }
+    uint64_t ret_key = MPIDI_OFI_INVALID_MR_KEY;
+    int id = key_type;
+
+    switch (key_type) {
+        case MPIDI_OFI_LOCAL_MR_KEY:
+            {
+                uint64_t i;
+                for (i = mr_key_allocator.last_free_mr_key; i < mr_key_allocator.num_ints; i++) {
+                    if (mr_key_allocator.bitmask[i]) {
+                        register uint64_t val, nval;
+                        val = mr_key_allocator.bitmask[i];
+                        nval = 2;
+                        MPIDI_OFI_INDEX_CALC(val, nval, 32, 0xFFFFFFFFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 16, 0xFFFFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 8, 0xFFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 4, 0xFULL);
+                        MPIDI_OFI_INDEX_CALC(val, nval, 2, 0x3ULL);
+                        nval -= val & 0x1ULL;
+                        mr_key_allocator.bitmask[i] &= ~(0x1ULL << (nval - 1));
+                        mr_key_allocator.last_free_mr_key = i;
+                        ret_key = i * sizeof(uint64_t) * 8 + (nval - 1);
+                    }
+                    if (i == mr_key_allocator.num_ints - 1) {
+                        mr_key_allocator.num_ints += mr_key_allocator.chunk_size;
+                        mr_key_allocator.bitmask = MPL_realloc(mr_key_allocator.bitmask,
+                                                               sizeof(uint64_t) *
+                                                               mr_key_allocator.num_ints,
+                                                               MPL_MEM_RMA);
+                        MPIR_Assert(mr_key_allocator.bitmask);
+                        memset(&mr_key_allocator.bitmask[i + 1], 0xFF,
+                               sizeof(uint64_t) * mr_key_allocator.chunk_size);
+                    }
+                }
+                break;
+            }
+
+        case MPIDI_OFI_COLL_MR_KEY:
+            {
+                MPIR_Assert(requested_key != MPIDI_OFI_INVALID_MR_KEY);
+                ret_key = requested_key;
+                break;
+            }
+
+        case default:
+            {
+                MPIR_Assert(0);
+            }
     }
-    return -1;
+
+    if (ret_key != MPIDI_OFI_INVALID_MR_KEY) {
+        ret_key = (ret_key << MPIDI_OFI_MR_KEY_PREFIX_SHIFT) | key_type;
```

Comment:
My proposal to use a shift operation here is more complicated than it needs to be. I think what we should do instead is just
```
if (coll) {
    ret_key = requested_key;
    ret_key |= 1ULL << 63;
}
```
basically we use the highest bit as the collective indicator. Assuming local keys are allocated in ascending order, this mean collisions won't happen unless they are a crazy number of local keys allocated.

Reply:
sounds good, I will change it